File: backend/src/main/java/com/chami/backend/services/PostService.java
package com.chami.backend.services;

import com.chami.backend.dtos.PostDTO;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

public interface PostService {
    PostDTO createPost(PostDTO postDTO, MultipartFile mediaFile);
    PostDTO updatePost(Long id, PostDTO postDTO, MultipartFile mediaFile);
    
    // New methods for direct URL handling
    PostDTO createPostWithDirectMedia(PostDTO postDTO);
    PostDTO updatePostWithDirectMedia(Long id, PostDTO postDTO);
    
    // Methods for likes and shares
    PostDTO likePost(Long postId, Long userId);
    PostDTO unlikePost(Long postId, Long userId);
    PostDTO incrementShareCount(Long postId);
    void deletePost(Long id);
    PostDTO getPostById(Long id);
    List<PostDTO> getAllPosts();
    List<PostDTO> getPostsByUser(Long userId);
    List<PostDTO> searchPosts(String keyword);
    List<PostDTO> getPostsByTag(String tag);
}

File: frontend/src/services/post.service.js
import api from './api';

class PostService {
  getAllPosts() {
    return api.get('/posts');
  }

  getPostById(id) {
    return api.get(`/posts/${id}`);
  }

  getPostsByUser(userId) {
    return api.get(`/posts/user/${userId}`);
  }

  createPost(postData) {
    return api.post("/posts", postData);
  }

  updatePost(id, postData) {
    return api.put(`/posts/${id}`, postData);
  }

  deletePost(id) {
    return api.delete(`/posts/${id}`);
  }

  likePost(id) {
    return api.post(`/posts/${id}/like`);
  }

  unlikePost(id) {
    return api.post(`/posts/${id}/unlike`);
  }

  sharePost(id) {
    return api.post(`/posts/${id}/share`);
  }

  getCommentsByPost(postId) {
    return api.get(`/comments/post/${postId}`);
  }

  addComment(postId, content) {
    return api.post('/comments', {
      postId: postId,
      content: content
    });
  }

  updateComment(id, content) {
    return api.put(`/comments/${id}`, { content });
  }

  deleteComment(id) {
    return api.delete(`/comments/${id}`);
  }

  searchPosts(keyword) {
    // Use encodeURIComponent to properly handle special characters in the search term
    return api.get(`/posts/search?keyword=${encodeURIComponent(keyword)}`);
  }

  getPostsByTag(tag) {
    return api.get(`/posts/tag/${tag}`);
  }
}

export default new PostService();
